docs/book/tutorial/page-3.md

```diff
@@ -0,0 +1,736 @@
+	}
+	job.Annotations[scheduledTimeAnnotation] = scheduledTime.Format(time.RFC3339)
+	for k, v := range cronJob.Spec.JobTemplate.Labels {
+		job.Labels[k] = v
+	}
+	if err := ctrl.SetControllerReference(cronJob, job, r.scheme); err != nil {
+		return nil, err
+	}
+
+	return job, nil
+}
+
+```
+<a name="jump-to-func--r--CronJobReconciler--ConstructJobForCronJob"></a>
+[return](#return-from-func (r *CronJobReconciler) ConstructJobForCronJob)
+%!(EXTRA string=func--r--CronJobReconciler--ConstructJobForCronJob){%% endmethod %%}
+
+{% method %}
+
+We'll use a helper to extract the scheduled time from the annotation that
+we added during job creation.
+
+
+```go
+
+func getScheduledTimeForJob(job *v1.Job) (*time.Time, error) {
+	timeRaw := job.Annotations[scheduledTimeAnnotation]
+	if len(timeRaw) == 0 {
+		return nil, nil
+	}
+
+	timeParsed, err := time.Parse(time.RFC3339, timeRaw)
+	if err != nil {
+		return nil, err
+	}
+	return &timeParsed, nil
+}
+
+```
+<a name="jump-to-func-getScheduledTimeForJob"></a>
+[return](#return-from-func getScheduledTimeForJob)
+%!(EXTRA string=func-getScheduledTimeForJob){%% endmethod %%}
+
+{% method %}
+
+We'll calculate the next scheduled time using our helpful cron library.
+We'll start calculating appropriate times from our last run, or the creation
+of the CronJob if we can't find a last run.
+
+If there are too many missed runs and we don't have any deadlines set, we'll
+bail so that we don't cause issues on controller restarts or wedges.
+
+Otherwise, we'll just return the missed runs (of which we'll just use the latest),
+and the next run, so that we can know the latest time to reconcile again.
+
+
+```go
+
+
+// getNextSchedule fetches any recently missed runs of this cronjob that we might
+// want to create a job for, as well as the next run we're expecting to make
+// (so that we can wait till then).
+func getNextSchedule(cronJob *v1.CronJob, now time.Time) ([]time.Time, time.Time, error) {
+	starts := []time.Time{}
+	sched, err := cron.ParseStandard(cronJob.Spec.Schedule)
+	if err != nil {
+		return nil, time.Time{}, fmt.Errorf("Unparseable schedule %q: %v", cronJob.Spec.Schedule, err)
+	}
+
+	var earliestTime time.Time
+	if cronJob.Status.LastScheduleTime != nil {
+		earliestTime = cronJob.Status.LastScheduleTime.Time
+	} else {
+		// If none found, then this is either a recently created scheduledJob,
+		// or the active/completed info was somehow lost (contract for status
+		// in kubernetes says it may need to be recreated), or that we have
+		// started a job, but have not noticed it yet (distributed systems can
+		// have arbitrary delays).  In any case, use the creation time of the
+		// CronJob as last known start time.
+		earliestTime = cronJob.ObjectMeta.CreationTimestamp.Time
+	}
+	if cronJob.Spec.StartingDeadlineSeconds != nil {
+		// Controller is not going to schedule anything below this point
+		schedulingDeadline := now.Add(-time.Second * time.Duration(*cronJob.Spec.StartingDeadlineSeconds))
```

Comment:
> now.Add(-time.Second * time.Duration(*cronJob.Spec.StartingDeadlineSeconds))

It seems it's wrong here. `now.Add(time.Duration(*cronJob.Spec.StartingDeadlineSeconds))` should do the job.